anticipate access token refresh

ironage opened the pull request:
This is to address: https://jira.mongodb.org/browse/RCORE-473
I added back the old WaitingForAccessToken state to the sync session, though the definition of an access token has changed a bit so the flow is much simpler.

Without the changes in this PR, I also did a manual test of logging in a user, then waiting more than 30 minutes to let the access token expire and then using it to start a sync session and the error handling and reconnection all worked for me. So the only thing I can claim to have fixed here for sure is that there are less errors logged (both client side and server side). However, the alarming (though harmless) error logs seems to be what users are concerned about (https://github.com/realm/realm-js/issues/3667).

That being said, the test added does trigger the new code, but will pass without these changes because the 401 response received when initiating a sync session with an expired access token does already get handled correctly (since https://github.com/realm/realm-object-store/pull/1024).

## ☑️ ToDos
* [x] 📝 Changelog update
* [x] 🚦 Tests (or not relevant)

tgoyne reviewed (commented):
Looking at the old version of the WaitingForAccessToken state, most of it does appear to be unneeded now but nonsync_transact_notify() is probably still needed to trigger an upload of any commits which happened while the session was waiting for the token.

The SyncServer test type injects a custom clock into the server which should be usable to test this more directly (connect to server, advance server's clock enough to make the token not longer valid, then verify that the refresh happens). I'm not sure if that code still works with v10, though? I suspect none of the tests related to access token expiration ever got ported over even though it's all still applicable.

ironage commented:
Thanks for the feedback, I've addressed the comments.
I doubt the access token tests were ported over, at least I can't recall doing that myself. But the definition of an access token has changed a bit as well, because now, an access token is valid for 30 minutes and I'm pretty sure it is only used to initiate a sync session and for API requests instead of being checked against every incoming sync change. 
For the test I added at least, the token is running against the docker MongoDB test instance so I don't think we can modify it's clock artificially in the same way.

fealebenpae reviewed (approved):
Might be an overkill to bring all that machinery back, but until we can refactor the transport layer in the sync client this is a good enough fix, considering the time we have 👍

simonask commented on test/object-store/util/test_utils.cpp:
```diff
@@ -70,12 +70,24 @@ void catch2_ensure_section_run_workaround(bool did_run_a_section, std::string se
+    // by default make a valid expiry time so that the sync session pre check
+    // doesn't trigger a token refresh on first open
+    using namespace std::chrono_literals;
+    if (!exp) {
```
Kind of confusing that `0` means "30 minutes from now", I guess? Consider changing `exp` and `iat` to be `Optional<int64_t>`.

simonask commented on CHANGELOG.md:
```diff
@@ -5,7 +5,7 @@
 ### Fixed
 * <How to hit and notice issue? what was the impact?> ([#????](https://github.com/realm/realm-core/issues/????), since v?.?.?)
-* None.
+* Proactivly check the expiry time on the access token and refresh it before attempting to initiate a sync session. This prevents some error logs from appearing on the client such as: "ERROR: Connection[1]: Websocket: Expected HTTP response 101 Switching Protocols, but received: HTTP/1.1 401 Unauthorized" ([RCORE-473](https://jira.mongodb.org/browse/RCORE-473), since v10.0.0)
```
"Proactively"

simonask reviewed (approved):
Looks good, just minor comments :+1:

tgoyne commented:
This is missing handling for the new state in `get_public_state()`, so the Cocoa tests hit the following:

```
/System/Volumes/Data/src/rc-master/.build/checkouts/realm-core/src/realm/object-store/sync/sync_session.cpp:931: [realm-core-10.6.1] Unreachable code
0   RealmPackageTests                   0x000000011031aafc _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 28
1   RealmPackageTests                   0x000000011031adc1 _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 257
2   RealmPackageTests                   0x000000010e4e7085 _ZNK5realm11SyncSession16get_public_stateEv + 213
3   RealmPackageTests                   0x000000010e4e6cc3 _ZN5realm11SyncSession23add_completion_callbackERKNSt3__111unique_lockINS1_5mutexEEENS1_8functionIFvNS1_10error_codeEEEENS_5_impl20SyncProgressNotifier12NotifierTypeE + 291
4   RealmPackageTests                   0x000000010e4e7469 _ZN5realm11SyncSession28wait_for_download_completionENSt3__18functionIFvNS1_10error_codeEEEE + 137
5   RealmPackageTests                   0x000000010e45c0b2 _ZN5realm13AsyncOpenTask5startENSt3__18functionIFvNS_19ThreadSafeReferenceESt13exception_ptrEEE + 258
6   RealmPackageTests                   0x000000010e800ca9 __62+[RLMRealm asyncOpenWithConfiguration:callbackQueue:callback:]_block_invoke + 425
7   libdispatch.dylib                   0x00007fff2020e5dd _dispatch_call_block_and_release + 12
8   libdispatch.dylib                   0x00007fff2020f7c7 _dispatch_client_callout + 8
9   libdispatch.dylib                   0x00007fff202155fe _dispatch_lane_serial_drain + 606
10  libdispatch.dylib                   0x00007fff202160cb _dispatch_lane_invoke + 375
11  libdispatch.dylib                   0x00007fff2021fc5d _dispatch_workloop_worker_thread + 819
12  libsystem_pthread.dylib             0x00007fff203b7499 _pthread_wqthread + 314
13  libsystem_pthread.dylib             0x00007fff203b6467 start_wqthread + 15!!! IMPORTANT: Please report this at https://github.com/realm/realm-core/issues/new/chooseExited with signal code 6
```
